zipline/pipeline/loaders/events.py

```diff
@@ -0,0 +1,154 @@
+import numpy as np
+import pandas as pd
+from six import iteritems
+from toolz import merge
+
+from .base import PipelineLoader
+from .frame import DataFrameLoader
+from .utils import next_date_frame, previous_date_frame, previous_value
+
+TS_FIELD_NAME = "timestamp"
+SID_FIELD_NAME = "sid"
+
+
+class EventsLoader(PipelineLoader):
+    """
+    Abstract loader.
+
+    Does not currently support adjustments to the dates of known events.
+
+    Parameters
+    ----------
+    all_dates : pd.DatetimeIndex
+        Index of dates for which we can serve queries.
+    events_by_sid : dict[int -> pd.DataFrame]
+        Dict mapping sids to DataFrames representing dates on which events
+        occurred along with other associated values.
+
+        If the DataFrames contain a "timestamp" column, that column is
+        interpreted as the date on which we learned about the event.
+
+        If the DataFrames do not contain a "timestamp" column, we assume we
+        knew about the event on all prior dates.  This mode is only supported
+        if ``infer_timestamp`` is explicitly passed as a truthy value.
+    infer_timestamps : bool, optional
+        Whether to allow omitting the "timestamp" column.
+    dataset : DataSet
+        The DataSet object for which this loader loads data.
+    expected_cols : frozenset
+        Set of expected columns for the dataset, without timestamp.
+    """
+
+    def __init__(self,
+                 all_dates,
+                 events_by_sid,
+                 infer_timestamps=False,
+                 dataset=None,
+                 expected_cols=frozenset()):
+        self.all_dates = all_dates
+        # Do not modify the original in place, since it may be used for other
+        #  purposes.
+        self.events_by_sid = (
+            events_by_sid.copy()
+        )
+        dates = self.all_dates.values
+
+        for k, v in iteritems(events_by_sid):
+            # First, must convert to DataFrame.
+            if isinstance(v, pd.Series):
+                # If Series was passed, DateTime index is assumed.
+                self.events_by_sid[k] = pd.DataFrame(v)
+            elif isinstance(v, pd.DatetimeIndex):
+                if not infer_timestamps:
+                    raise ValueError(
+                        "Got DatetimeIndex for sid %d.\n"
+                        "Pass `infer_timestamps=True` to use the first date in"
+                        " `all_dates` as implicit timestamp." % k
+                    )
+                self.events_by_sid[k] = pd.DataFrame(v)
+                v.index = [dates[0]] * len(v)
+            # Already a DataFrame
+            elif isinstance(v, pd.DataFrame):
```

Comment:
If we hit either of the two previous branches, we won't need to make this check, because there will never be a `TS_FIELD_NAME` column at this point, though the index will already have been set. So I don't think it would make sense for the DataFrames generated in the other two cases to go through these checks, hence the `elif`.

What do you mean we'll silently ignore anything that's not any of those 3? That's why I raise a `ValueError` if we don't enter any of those 3 previous cases.

Yeah, I think I'll need to add tests in `test_events.py`.